File: src/main/java/org/goafabric/personservice/persistence/extensions/DemoDataImporter.java
/*
package org.goafabric.personservice.persistence.extensions;

import jakarta.enterprise.context.ApplicationScoped;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.goafabric.personservice.controller.dto.Address;
import org.goafabric.personservice.controller.dto.Person;
import org.goafabric.personservice.extensions.TenantContext;
import org.goafabric.personservice.logic.PersonLogic;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

@ApplicationScoped
public class DemoDataImporter {
    private final Logger log = LoggerFactory.getLogger(this.getClass());

    private final String goals;

    private final String tenants;

    private final PersonLogic personLogic;


    public DemoDataImporter(@ConfigProperty(name = "database.provisioning.goals") String goals, @ConfigProperty(name = "multi-tenancy.tenants")String tenants, PersonLogic personLogic) {
        this.goals = goals;
        this.tenants = tenants;
        this.personLogic = personLogic;
    }

    public void run() {
        if (goals.contains("-import-demo-data")) {
            log.info("Importing demo data ...");
            importDemoData();
            log.info("Demo data import done ...");
        }

        if (goals.contains("-terminate")) {
            log.info("Terminating app ...");
            System.exit(0);
        }
    }

    private void importDemoData() {
        Arrays.asList(tenants.split(",")).forEach(tenant -> {
            TenantContext.setTenantId(tenant);
            if (personLogic.findAll().isEmpty()) {
                insertData();
            }
        });
        TenantContext.setTenantId("0");
    }

    private void insertData() {
        IntStream.range(0, 1).forEach(i -> {
            personLogic.save(new Person(null, null, "Homer", "Simpson"
                    , List.of(createAddress("Evergreen Terrace No. " + i))));

            personLogic.save(new Person(null, null, "Bart", "Simpson"
                    , List.of(createAddress("Everblue Terrace No. " + i))));

            personLogic.save(new Person(null, null, "Monty", "Burns"
                    , List.of(createAddress("Mammon Street No. 1000 on the corner of Croesus"))));
        });

    }

    private Address createAddress(String street) {
        return new Address(null, null, street, "Springfield " + TenantContext.getTenantId());
    }

}


 */

File: src/main/kotlin/org/goafabric/personservice/persistence/extensions/DemoDataImporter.kt
package org.goafabric.personservice.persistence.extensions

import io.quarkus.runtime.StartupEvent
import jakarta.enterprise.context.ApplicationScoped
import jakarta.enterprise.event.Observes
import org.eclipse.microprofile.config.inject.ConfigProperty
import org.goafabric.personservice.controller.dto.Address
import org.goafabric.personservice.controller.dto.Person
import org.goafabric.personservice.controller.dto.PersonSearch
import org.goafabric.personservice.extensions.UserContext
import org.goafabric.personservice.logic.PersonLogic
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.function.Consumer
import java.util.function.IntConsumer
import java.util.stream.IntStream

@ApplicationScoped
class DemoDataImporter(
    @param:ConfigProperty(name = "database.provisioning.goals") private val goals: String,
    @param:ConfigProperty(name = "multi-tenancy.tenants") private val tenants: String,
    private val personLogic: PersonLogic
) {
    private val log: Logger = LoggerFactory.getLogger(this.javaClass)

    fun onStart(@Observes ev: StartupEvent) {
        run()
    }

    fun run() {
        if (goals.contains("-import-demo-data")) {
            log.info("Importing demo data ...")
            importDemoData()
            log.info("Demo data import done ...")
        }

        if (goals.contains("-terminate")) {
            log.info("Terminating app ...")
            System.exit(0)
        }
    }

    private fun importDemoData() {
        listOf(*tenants.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()).forEach(
            Consumer { tenant: String ->
                UserContext.tenantId = tenant
                if (personLogic.search(PersonSearch(null, null), 1, 10).isEmpty()) {
                    insertData()
                }
            })
        UserContext.tenantId = "0"
    }

    private fun insertData() {
        IntStream.range(0, 1).forEach(IntConsumer { i: Int ->
            personLogic.save(
                Person(
                    null, null, "Homer", "Simpson",
                    listOf<Address>(createAddress("Evergreen Terrace No. " + i))
                )
            )
            personLogic.save(
                Person(
                    null, null, "Bart", "Simpson",
                    listOf<Address>(createAddress("Everblue Terrace No. " + i))
                )
            )
            personLogic.save(
                Person(
                    null, null, "Monty", "Burns",
                    listOf<Address>(createAddress("Mammon Street No. 1000 on the corner of Croesus"))
                )
            )
        })
    }

    private fun createAddress(street: String): Address {
        return Address(null, null, street, "Springfield " + UserContext.tenantId)
    }
}

